### ISSUE TITLE
Problem archiving app with use_frameworks!

### ISSUE THREAD

**Issue by choefele** [CocoaPods/CocoaPods#3493]
A CocoaPods project with multiple targets fails archiving with errors. Please see below on how to reproduce.

Xcode 6.3.1
CocoaPods 0.37.0

Here are the steps to reproduce the problem:
- Create new project with Xcode (single view, 'Test')
- Add Today Extension ('Widget')
- Use below Podfile
- Run `pod install`
- Delete derived data to make sure there are no left-overs
- Open workspace in Xcode and select iOS Device
- Run Product > Archive

```Ruby
platform :ios, '8.0'
use_frameworks!

source 'https://github.com/CocoaPods/Specs'

link_with 'Widget'

pod 'CCHBinaryData'

target 'Test' do
    pod 'TSMessages'
end
```

The archive operation succeeds the first time because the derived data was deleted before archiving. However, the app doesn't appear in the Archives tool (Window > Organizer).

- Run Product > Archive again

The archive operation fails with the following error messages:

```
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/lipo: can't move temporary file: /Users/choefele/Library/Developer/Xcode/DerivedData/Test-aprwsydqlazvlbacurkmubqnhure/Build/Intermediates/ArchiveIntermediates/Test/IntermediateBuildFilesPath/UninstalledProducts/CCHBinaryData.framework/CCHBinaryData to file: /Users/choefele/Library/Developer/Xcode/DerivedData/Test-aprwsydqlazvlbacurkmubqnhure/Build/Intermediates/ArchiveIntermediates/Test/IntermediateBuildFilesPath/UninstalledProducts/CCHBinaryData.framework/CCHBinaryData.lipo (No such file or directory)

Command /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/lipo emitted errors but did not return a nonzero exit code to indicate failure

Command /usr/bin/codesign failed with exit code 1
``` 

Full log output: https://gist.github.com/choefele/e820ad2194c7539b3e84

![bildschirmfoto 2015-05-04 um 11 03 13](https://cloud.githubusercontent.com/assets/3195446/7451121/3126596a-f24d-11e4-91cd-fb7d86abebd5.png)

**Comment by mrackwitz** [comment]
Thanks for the report, @choefele! This seems to be a duplicate of #3156, which is in general about a code signing issue, but describes also related problems while archiving. This https://github.com/CocoaPods/CocoaPods/issues/3156#issuecomment-88454250 gives some hints about the underlying issue with Xcode's parallelized builds. You can turn them off with:
>Product > Scheme > Edit Scheme > Select Build on the left > Build Options > Parallelize Build Off

**Comment by choefele** [comment]
Setting Parallelize Build to Off makes the archive operation succeed, but the resulting archive still doesn't turn up in the Organizer.

I think the underlying problem here is that cocoapods creates two targets for the same framework (Pods-CCHBinaryData + Pods-Test-CCHBinaryData). The resulting app however, should only contain one framework that both targets use. Can these targets be merged together? I don't think it's safe to build the same framework twice with slightly different parameters (e.g. Allow app extension only).

**Comment by mrackwitz** [comment]
#### Known Limitations:

>Setting Parallelize Build to Off makes the archive operation succeed, but the resulting archive still doesn't turn up in the Organizer.
>
>I think the underlying problem here is that cocoapods creates two targets for the same framework (Pods-CCHBinaryData + Pods-Test-CCHBinaryData). The resulting app however, should only contain one framework that both targets use. Can these targets be merged together?

They could be merged, theoretically. That's a known issue, which came up internally and in the discussion here #3095 and is blocked by the new Podfile DSL (#840). But this shouldn't cause the issue, which appears for you with the missing build products in the Organizer.

>I don't think it's safe to build the same framework twice with slightly different parameters (e.g. Allow app extension only).

As those are namespaced in different `CONTENTS_FOLDER_PATH`s (by the target where they are integrated), it's actually safe to build them. It wouldn't be even unsafe to run them, as extensions run in a separate process, they don't conflict from linking at runtime with the main app or with each other. What doesn't work indeed is to ship a separate sets of frameworks in an `.appex` bundle to the AppStore, that is rejected in the prevalidation phase as `Invalid bundle`. This is addressed by #3489.


#### Further troubleshooting:

Doesn't appear any new product in the organizer after archiving? Are you sure that there isn't another target in your scheme, which is built in the same directory and causes that Xcode is creating an `.xcarchive` instead? Did you double-checked the options `SKIP_INSTALL`, `INSTALL_PATH` and manually checked the contents of `ARCHIVE_PATH`?


#### Workaround:

If you want to solve that, the only solution I see for now, is to add a post install hook to your Podfile, how @tannernelson did it in https://github.com/CocoaPods/CocoaPods/issues/3156#issuecomment-90964547:

```ruby
post_install do |installer_representation|
    installer_representation.project.targets.each do |target|
        # Matches all Pods which are part of the default aggregate target (Pods),
        # which is linked to the first target in your project, which is very likely your
        # app target.
        if target.to_s.match /Pods-[^-]+$/
            puts target
            target.build_configurations.each do |config|
                config.build_settings['APPLICATION_EXTENSION_API_ONLY'] = 'YES'
            end
        end
    end
end
```

The pod targets, which are matched by the script above are those, whose build products are integrated in your app's `FRAMEWORKS_FOLDER_PATH` and are linked at runtime by your app and your extensions. That another framework build was linked at compile time to your extension shouldn't matter as they are the same (as long as you don't use different subspec sets, which you should avoid).

**Comment by choefele** [comment]
I created this project from scratch with the steps outlined in the description of this issue. Thus I'm 100% certain there are no settings in addition to the defaults.

Test project is available here: https://www.dropbox.com/s/89whnlxm0ju7pcb/Test.zip?dl=0 (This is after disabling parallelized builds in the scheme. Archiving works, but the resulting app doesn't show up in the Organizer.)

- `SKIP_INSTALL` is set to `NO` for the app target
- `INSTALL_PATH` is set to `/Application` – that seems to be the correct coniguration
- I can't find `ARCHIVE_PATH` anywhere

The last message when archiving refers to `/Users/choefele/Library/Developer/Xcode/DerivedData/Test-aprwsydqlazvlbacurkmubqnhure/Build/Intermediates/ArchiveIntermediates/Test/InstallationBuildProductsLocation/Applications/Test.app`, which I can't find at all. There's no app or xcarchive anywhere in the derived data directory.

**Comment by mrackwitz** [comment]
I can confirm the results which you've reported.
@neonichu: Any idea what could be wrong here and cause this problem?

**Comment by ilessiivi** [comment]
I'm struggling with this issue as well. When archiving, I get this message in my Mac's Console:

Xcode[27942]: Cannot make directory /Users/[my user]/Library/Developer/Xcode/Archives/2015-05-07/Test 7.5.2015 11.14.xcarchive/dSYMs/CCHBinaryData.framework.dSYM: File exists

Apparently it tries to generate the framework dSYM file twice but the second time fails and it stops the archiving process?

**Comment by SlaunchaMan** [comment]
Seeing this here, too.

**Comment by neonichu** [comment]
@mrackwitz my guess is this is a variant of #3240, caused by archiving multiple frameworks with the same module name.

**Comment by SlaunchaMan** [comment]
To give more detail:

I have a project with an iOS app and a framework. Both the iOS app and the framework link to Alamofire. The Podfile was set up like this:
```
platform :ios, '8.0'

source 'https://github.com/CocoaPods/Specs.git'

use_frameworks!

pod 'Alamofire'

link_with 'AppName', 'AppName Tests'

target :'AppName UI Tests', :exclusive => true do
    pod 'KIF'
end

target :FrameworkName, :exclusive => true do
    pod 'Alamofire'
end

target :FrameworkNameTests, :exclusive => true do
    pod 'Alamofire'
end
```

To fix it, I simply used `link_with` to link with the framework as well:
```
platform :ios, '8.0'

source 'https://github.com/CocoaPods/Specs.git'

use_frameworks!

pod 'Alamofire'

link_with 'AppName', 'AppName Tests', 'FrameworkName'

target :'AppName UI Tests', :exclusive => true do
    pod 'KIF'
end

target :FrameworkNameTests, :exclusive => true do
    pod 'Alamofire'
end
```

Then my archive worked fine.

**Comment by mrackwitz** [comment]
There's work going on to fix this in #3550 - but it's not quite ready, yet.

**Comment by allanmacatingrao** [comment]
Additional Issue:

How could I create a mix output using use_frameworks! and static lib
Example:


target :'MyApp' do
use_frameworks!
pod 'AFNetworking'

#don't use framework in succeeding pod
pod ...
pod ...
pod ..
end

Thanks.

**Comment by mrackwitz** [comment]
@allanmacatingrao: That's not possible. Frameworks are currently an all-or-nothing approach, because of the implications of dynamic vs. static linking.

**Comment by segiddins** [comment]
This should be fixed now that #3550 is done.

**Comment by Ying1986** [comment]
cocoa pod: 0.39
Xcode 7.2.1:

![podsetting](https://cloud.githubusercontent.com/assets/6648212/12910468/a1be6a48-cf13-11e5-9e39-4598dbabc028.png)
![bitcode-compile](https://cloud.githubusercontent.com/assets/6648212/12910464/a0c12ae0-cf13-11e5-9848-999c645c7b61.png)
![ipaerror-1](https://cloud.githubusercontent.com/assets/6648212/12910466/a1b94f04-cf13-11e5-9930-b5394c5f8731.png)
![ipaerror-2](https://cloud.githubusercontent.com/assets/6648212/12910467/a1bb71da-cf13-11e5-9e80-10e207dd67df.png)
